File: Irreducible Sum of Rationals/irreducible-sum-of-rationals.py
from functools import reduce
from math import gcd


def lcms(numbers):
    """
    source for lcms func - https://gist.github.com/endolith/114336/eff2dc13535f139d0d6a2db68597fad2826b53c3
    :type numbers: list
    :rtype: int
    """
    return int(reduce(lambda x, y: (x * y) / gcd(int(x), int(y)), numbers, 1))


def sum_fracts(lst):
    """
    :type lst: list
    :rtype: list
    """
    if not len(lst):
        return None

    numerator = 0
    denominator = lcms([x[1] for x in lst])

    for i in range(0, len(lst)):
        k = int(denominator / lst[i][1])
        numerator += k * lst[i][0]

    final_gcd = gcd(numerator, denominator)
    return [int(numerator / final_gcd), int(denominator / final_gcd)] \
        if numerator % denominator != 0 else int(numerator / denominator)


print(sum_fracts([[62, 63], [62, 64]]))  # [3937, 2016]
print(sum_fracts([[1, 3], [5, 3]]))  # 2
print(sum_fracts([[1, 2], [1, 3], [1, 4]]))  # [13, 12]

File: Irreducible Sum of Rationals/irreducible-sum-of-rationals.js
Math.gcd = (a, b) => {
  if ( ! b) {
    return a;
  }

  return Math.gcd(b, a % b);
};

/**
 * Calculate lcms
 *
 * @param {array} n
 * @return {number}
 */
const lcms = n => {
  return n.reduce((x, y) => parseInt(((x * y) / Math.gcd(x, y))), 1);
};

/**
 * Complete the sumFracts function below.
 *
 * @param {array} l
 * @return {array}|{number}|{null}
 */
const sumFracts = l => {
  if (!l.length) {
    return null;
  }

  let numerator = 0;
  const denominator = lcms(l.map(n => n[1]));

  let k;
  for (let i = 0; i < l.length; i++) {
    k = parseInt(denominator / l[i][1]);
    numerator += k * l[i][0];
  }

  const final_gcd = Math.gcd(numerator, denominator);

  return (numerator % denominator !== 0) ?
    [parseInt(numerator / final_gcd), parseInt(denominator / final_gcd)] : parseInt(numerator / denominator);
};

console.log(sumFracts([[62, 63], [62, 64]])); // [3937, 2016]
console.log(sumFracts([[1, 3], [5, 3]])); // 2
console.log(sumFracts([[1, 2], [1, 3], [1, 4]])); // [13, 12]